### PR TITLE
[SPARK-18191][CORE] Port RDD API to use commit protocol

### PR DESCRIPTION
**Pull request by jiangxb1987**
## What changes were proposed in this pull request?

This PR port RDD API to use commit protocol, the changes made here:
1. Add new internal helper class that saves an RDD using a Hadoop OutputFormat named `SparkNewHadoopWriter`, it's similar with `SparkHadoopWriter` but uses commit protocol. This class supports the newer `mapreduce` API, instead of the old `mapred` API which is supported by `SparkHadoopWriter`;
2. Rewrite `PairRDDFunctions.saveAsNewAPIHadoopDataset` function, so it uses commit protocol now.

## How was this patch tested?
Exsiting test cases.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/SparkHadoopWriter.scala [hunk 1]
```diff
@@ -162,11 +162,14 @@ class SparkHadoopWriter(jobConf: JobConf) extends Logging with Serializable {
...
-    val formatter = new SimpleDateFormat("yyyyMMddHHmmss", Locale.US)
-    val jobtrackerID = formatter.format(time)
+    val jobtrackerID = createJobTrackerID(time)
     new JobID(jobtrackerID, id)
   }
 
+  def createJobTrackerID(time: Date): String = {
+    new SimpleDateFormat("yyyyMMddHHmmss", Locale.US).format(time)
+  }
+
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/SparkHadoopWriter.scala]: We need to generate jobTrackerID seprately in `SparkNewHadoopWriter `

File: core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala [hunk 2]
```diff
@@ -509,21 +509,6 @@ class PairRDDFunctionsSuite extends SparkFunSuite with SharedSparkContext {
       (2, ArrayBuffer(1))))
   }
 
-  test("saveNewAPIHadoopFile should call setConf if format is configurable") {
```
Comment (jiangxb1987) [inline core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala]: This logic is no longer needed.

File: core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala [hunk 3]
```diff
@@ -725,8 +710,7 @@ class PairRDDFunctionsSuite extends SparkFunSuite with SharedSparkContext {
 }
 
 /*
-  These classes are fakes for testing
-    "saveNewAPIHadoopFile should call setConf if format is configurable".
+  These classes are fakes for testing saveAsHadoopFile/saveNewAPIHadoopFile.
```
Comment (jiangxb1987) [inline core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala]: This comment have been out-of-date for a while.

**Review by jiangxb1987 — COMMENTED** [review]
I didn't touch `SparkHadoopWriter` and `PairRDDFunctions.saveAsHadoopDataset`, because they use the older `mapred` API, which is not supported by the `FileCommitProtocol` framework now.

File: core/src/main/scala/org/apache/spark/SparkNewHadoopWriter.scala [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+ * (from the newer mapreduce API, not the old mapred API).
+ *
+ * Saves the RDD using a JobConf, which should contain an output key class, an output value class,
+ * a filename to write to, etc, exactly like in a Hadoop MapReduce job.
+ *
+ * Use a [[HadoopMapReduceCommitProtocol]] to handle output commit, which, unlike Hadoop's
+ * OutputCommitter, is serializable.
+ */
+private[spark]
+class SparkNewHadoopWriter(
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/SparkNewHadoopWriter.scala]: move this into internal/io

**Comment by rxin** [comment]
Doesn't need to be in this PR, but can you also implement a HadoopMapRedCommitProtocol that supports the older mapred package's commiter?

**Comment by jiangxb1987** [comment]
Moved the logic from `saveAsNewAPIHadoopDataset` into `SparkNewHadoopWriter`, also added `SparkNewHadoopWriterUtils` to store functions used by Hadoop/NewHadoop writer.
In a followup PR, I'll implement a similar protocol to support the older mapred package's committer.

File: core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala [hunk 5]
```diff
@@ -1092,37 +1093,38 @@ class PairRDDFunctions[K, V](self: RDD[(K, V)])
       jobFormat.checkOutputSpecs(job)
     }
 
+    // Instantiate writer
+    val committer = FileCommitProtocol.instantiate(
+      className = classOf[HadoopMapReduceCommitProtocol].getName,
+      jobId = stageId.toString,
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala]: In fact I'm not sure what value should be assigned to `jobId` here, should it be `jobTrackerId` combined with `jobId`, or any value else? I failed to find some code to follow on this topic.

File: core/src/main/scala/org/apache/spark/SparkHadoopWriter.scala [hunk 6]
```diff
@@ -23,11 +23,11 @@ import java.text.SimpleDateFormat
 import java.util.{Date, Locale}
 
 import org.apache.hadoop.fs.FileSystem
-import org.apache.hadoop.fs.Path
 import org.apache.hadoop.mapred._
 import org.apache.hadoop.mapreduce.TaskType
 
 import org.apache.spark.internal.Logging
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/SparkHadoopWriter.scala]: i'd move this file into spark/internal/io as well to be closer to SparkNewHadoopWriter.scala

Comment (rxin) [inline core/src/main/scala/org/apache/spark/SparkHadoopWriter.scala]: we can do it in a separate pr when you do the mapred committer.

File: core/src/main/scala/org/apache/spark/internal/io/SparkNewHadoopWriter.scala [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+
+/**
+ * A helper object that saves an RDD using a Hadoop OutputFormat
+ * (from the newer mapreduce API, not the old mapred API).
+ */
+private[spark]
+object SparkNewHadoopWriter extends Logging {
+
+  /** A shared job description for all the write tasks. */
+  private class WriteJobDescription[K, V](
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/internal/io/SparkNewHadoopWriter.scala]: maybe just remove this, since you have only 3 items here. In SQL there were a lot of items.

File: core/src/main/scala/org/apache/spark/internal/io/SparkNewHadoopWriter.scala [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+   *    rows within an RDD partition.
+   * 3. If no exception is thrown in a task, commits that task, otherwise aborts that task;  If any
+   *    exception is thrown during task commitment, also aborts that task.
+   * 4. If all tasks are committed, commit the job, otherwise aborts the job;  If any exception is
+   *    thrown during job commitment, also aborts the job.
+   */
+  def write[K, V: ClassTag](
+      sparkContext: SparkContext,
+      rdd: RDD[(K, V)],
+      committer: HadoopMapReduceCommitProtocol,
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/internal/io/SparkNewHadoopWriter.scala]: i'd move the creation of the commit protocol here here. The reason I put it outside in SQL was because streaming and batch needed to specify different protocols, but that problem doesn't exist in core.

File: core/src/main/scala/org/apache/spark/SparkHadoopWriter.scala (hunk 6 above)
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/SparkHadoopWriter.scala]: Yes I'll do it later.

**Comment by jiangxb1987** [comment]
The failed test case is imported by #15725, which is not related to our changes in this PR.

File: core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+        s"$committer may be an output committer that writes data directly to " +
+          "the final location. Because speculation is enabled, this output committer may " +
+          "cause data loss (see the case in SPARK-10063). If possible, please use an output " +
+          "committer that does not have this behavior (e.g. FileOutputCommitter)."
+      logWarning(warningMessage)
+    }
+
+    // Try to write all RDD partitions as a Hadoop OutputFormat.
+    try {
+      sparkContext.runJob(rdd, (context: TaskContext, iter: Iterator[(K, V)]) => {
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala]: we need to collect the result coming from the commit protocol here, and pass it into commitJob, don't we?

**Comment by rxin** [comment]
The rest looks good.

cc @ericl for another look too

**Comment by rxin** [comment]
Also @mridulm who has looked at one of the prs for commit protocol too.

@mridulm can you help review this too?

**Comment by jiangxb1987** [comment]
retest this please.

**Comment by jiangxb1987** [comment]
@rxin I've addressed your comment. Seems the test case is still failing... @ericl

**Comment by rxin** [comment]
I've disabled the test.

File: core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+        logError(s"Task ${taskContext.getTaskAttemptID} aborted.")
+      }, finallyBlock = writer.close(taskContext))
+    } catch {
+      case t: Throwable =>
+        throw new SparkException("Task failed while writing rows", t)
+    } finally {
+      outputMetricsAndBytesWrittenCallback.foreach { case (om, callback) =>
+        om.setBytesWritten(callback())
+        om.setRecordsWritten(recordsWritten)
+      }
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala]: This looks like a behavior change - metric's are getting updated even when exceptions are thrown.
Do it after Utils.tryWithSafeFinallyAndFailureCallbacks completes, not in finally

File: core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+
+    // Set up a job.
+    val jobTrackerId = SparkHadoopWriterUtils.createJobTrackerID(new Date())
+    val jobAttemptId = new TaskAttemptID(jobTrackerId, stageId, TaskType.MAP, 0, 0)
+    val jobContext = new TaskAttemptContextImpl(conf.value, jobAttemptId)
+    val format = jobContext.getOutputFormatClass
+
+    if (SparkHadoopWriterUtils.isOutputSpecValidationEnabled(sparkConf)) {
+      // FileOutputFormat ignores the filesystem parameter
+      val jobFormat = format.newInstance
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala]: If it is Configurable, we should invoke setConf on it - the earlier code was doing this.
The tests for that seem to have been removed as well.
Any reason why ?

File: core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+      recordsWritten: Long): Unit = {
+    if (recordsWritten % RECORDS_BETWEEN_BYTES_WRITTEN_METRIC_UPDATES == 0) {
+      outputMetricsAndBytesWrittenCallback.foreach { case (om, callback) =>
+        om.setBytesWritten(callback())
+        om.setRecordsWritten(recordsWritten)
+      }
+    }
+  }
+
+  val RECORDS_BETWEEN_BYTES_WRITTEN_METRIC_UPDATES = 256
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala]: private and move it to top of the object ?

**Review by mridulm — COMMENTED** [review]
Overall looks good, added a few comments though.

File: core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala (hunk 11 above)
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala]: The logic is to create a `OutputCommitter` from the given hadoop conf, which is now handled by `HadoopMapReduceCommitProtocol`, so this code and corresponding tests are no longer needed.

**Comment by jiangxb1987** [comment]
@mridulm I've added most of your comments. Thank you!

**Comment by rxin** [comment]
Thanks - merging in master.

File: core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala (hunk 11 above)
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala]: I think I added the comment to the wrong location (though it is relevant here, it is probably less serious ?).
HadoopMapReduceCommitProtocol.setupCommitter() should be doing a setConf if OutputFormat is Configurable.

This needs to be fixed to ensure custom OutputFormat's work.
I see that the PR has already been committed - can you please file a bug and fix it ?
The test will also need to be re-added.

Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala]: Oh...I see the problem now. Will add that in a follow up, thank you for clarifying.

File: core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+   *    exception is thrown during task commitment, also aborts that task.
+   * 4. If all tasks are committed, commit the job, otherwise aborts the job;  If any exception is
+   *    thrown during job commitment, also aborts the job.
+   */
+  def write[K, V: ClassTag](
+      rdd: RDD[(K, V)],
+      hadoopConf: Configuration): Unit = {
+    // Extract context and configuration from RDD.
+    val sparkContext = rdd.context
+    val stageId = rdd.id
```
Comment (rezasafi) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala]: Is this accurate? Seems weird that satgeId is set to be equal to rdd.id. What is the commit protocol here?

Comment (jiangxb1987) [reply core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala]: It follows the previous behavior, what's your concern here?

Comment (rezasafi) [reply core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala]: We had a test failure using Spark 2.2 that seems to happen after this commit. The failure didn't occur before using Spark 2.0. We thought there could be a problem since the jobContext passed to commiter.setupJob() (line 84) has a different JobID comparing to the task context that is passed to to commiter.setupTask() (line 126)  in the former, it comes from rdd.id and in the later, its from stage.id. Just wanted to check what is the protocol here that requires such a difference? Thanks.

Comment (rezasafi) [reply core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala]: I investigated the problem a little more and I filed a jira & fix, see SPARK-22162. Thank you in advance.
